src/main/java/com/redhat/labs/lodestar/service/EngagementService.java

```diff
@@ -169,6 +173,30 @@ public Engagement update(String customerName, String projectName, Engagement eng
 
     }
 
+    /**
```

Comment:
I would not write this this way Instead -> 

`boolean doesSubdomainExist(String newSubdomain)` Just see if the subdomain exists when calling a create.
Don't set empty strings but handle nulls instead. If not set it should be null. (if you were checking for an empty string it should be "string".isEmpty().

in the create() method I would check if(false) then throw a new SubdomainExistsException() or InvalidEngagmentException or maybe reuse the InvalidRequestException. Specify the error message.

in the EngagementResource.post method - catch that error and return a bad request 400 with the error message. Maybe a code? if it helps the FE.

Also - add a new HEAD method `uniqueSubdomain` that accepts the subdomain and checks if it exist. 200 if available and 40x if taken (406?). Call directly to the doesSubdomainExist then in the FE you can check when the value is set by default and if the user changes before create.


I would just remove the validate all together for the update. Assume positive intent for the time being since it is not editable by the js.

Reply:
On reflection - the subdomain will be set after creation in the `update` method - I would check that there is a change from null to something and then check if it already exists. I believe on create it will not yet be set.

```
if(persisted == null && new != null && doesSubdomainExist(new) )
   ==> fail since it's trying to be set but is already taken
```

Reply:
Okay, I think I've hit all of your points, with just a few tweaks. Right now, the code is using exceptionmappers to map exceptions thrown in the service to response codes in the resource. Instead of reusing the `InvalidRequestException`, I used the `ResourceAlreadyExistsException`, as there is a `ResourceAlreadyExistsExceptionMapper` that handles mapping from that exception to a `409` status code. I created the `HEAD` request at `/engagements/subdomain/{subdomain}` that returns a `200` if available, and a `409` otherwise. I run the subdomain check in both the `create` and `update` of the `EngagementService`, and i moved from setting empty strings to null values, and then checking those null values in the respective methods.